graphql/src/execution/cache.rs

```diff
@@ -0,0 +1,240 @@
+            .set(value)
+            .unwrap_or_else(|_| panic!("Cache set should only be called once"));
+        // Wakeup consumers of the cache
+        let mut is_set = self.lock.lock().unwrap();
+        *is_set = true;
+        self.condvar.notify_all();
+    }
+
+    fn set(&self, value: R) {
+        self.set_inner(Some(value));
+    }
+
+    fn set_panic(&self) {
+        self.set_inner(None);
+    }
+
+    fn wait(&self) -> &R {
+        // Happy path - already cached.
+        if let Some(r) = self.result.get() {
+            match r.as_ref() {
+                Some(r) => r,
+                // TODO: Instead of having an Option,
+                // retain panic information and propagate it.
+                None => panic!("Query panicked"),
+            }
+        } else {
+            // Wait for the item to be placed in the cache.
+            let mut is_set = self.lock.lock().unwrap();
+            while !*is_set {
+                is_set = self.condvar.wait(is_set).unwrap();
+            }
+
+            self.wait()
+        }
+    }
+}
+
+/// One the last reference is removed, schedule for cleanup in the cache.
+impl<R> Drop for CacheEntryInner<R> {
+    fn drop(&mut self) {
+        self.cleanup.push(self.hash);
+    }
+}
+
+/// On drop, call set_panic on self.value,
+/// unless set was called.
+struct PanicHelper<R> {
+    value: Option<Arc<CacheEntryInner<R>>>,
+}
+
+impl<R> Drop for PanicHelper<R> {
+    fn drop(&mut self) {
+        if let Some(inner) = self.value.take() {
+            inner.set_panic();
+        }
+    }
+}
+
+impl<R> PanicHelper<R> {
+    fn new(value: Arc<CacheEntryInner<R>>) -> Self {
+        Self { value: Some(value) }
+    }
+    fn set(mut self, r: R) -> Arc<CacheEntryInner<R>> {
+        let value = self.value.take().unwrap();
+        value.set(r);
+        value
+    }
+}
+
+/// Cache that keeps a result around as long as it is still in use somewhere.
+/// The cache ensures that the query is not re-entrant, so multiple consumers
+/// of identical queries will not execute them in parallel.
+///
+/// This has a lot in common with AsyncCache in the network-services repo,
+/// but is sync instead of async, and more specialized.
+pub struct QueryCache<R> {
+    cleanup: CleanupQueue,
+    cache: Arc<Mutex<HashMap<Hash, Weak<CacheEntryInner<R>>>>>,
+}
+
+impl<R> QueryCache<R> {
+    pub fn new() -> Self {
+        Self {
+            cleanup: CleanupQueue::default(),
+            cache: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+    /// Assumption: Whatever F is passed in consistently returns the same
+    /// value for any input - for all values of F used with this Cache.
+    pub fn cached_query<F: FnOnce() -> R>(&self, hash: Hash, f: F) -> CachedResponse<R> {
+        // This holds it's own lock so make sure that this happens outside of
+        // holding any other lock.
+        let cleanup = self.cleanup.take();
+
+        let mut cache = self.cache.lock().unwrap();
```

Comment:
We should release this lock before running `f`!

Reply:
Excellent catch!